plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1236,39 +1235,49 @@ public void executeRemoveOrphanFiles(ConnectorSession session, IcebergTableExecu
 
         long expireTimestampMillis = session.getStart().toEpochMilli() - retention.toMillis();
         removeOrphanFiles(table, session, executeHandle.getSchemaTableName(), expireTimestampMillis);
-        removeOrphanMetadataFiles(table, session, executeHandle.getSchemaTableName(), expireTimestampMillis);
     }
 
     private void removeOrphanFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp)
     {
-        Set<String> validDataFilePaths = stream(table.snapshots())
-                .map(Snapshot::snapshotId)
-                .flatMap(snapshotId -> stream(table.newScan().useSnapshot(snapshotId).planFiles()))
-                .map(fileScanTask -> fileName(fileScanTask.file().path().toString()))
-                .collect(toImmutableSet());
-        Set<String> validDeleteFilePaths = stream(table.snapshots())
-                .map(Snapshot::snapshotId)
-                .flatMap(snapshotId -> stream(table.newScan().useSnapshot(snapshotId).planFiles()))
-                .flatMap(fileScanTask -> fileScanTask.deletes().stream().map(file -> fileName(file.path().toString())))
-                .collect(Collectors.toUnmodifiableSet());
-        scanAndDeleteInvalidFiles(table, session, schemaTableName, expireTimestamp, union(validDataFilePaths, validDeleteFilePaths), "/data");
+        Set<String> validMetadataFilePaths = new HashSet<>();
+        ImmutableSet.Builder<String> validDataFilePaths = ImmutableSet.builder();
+        ImmutableSet.Builder<String> validDeleteFilePaths = ImmutableSet.builder();
+
+        for (Snapshot snapshot : table.snapshots()) {
+            if (snapshot.manifestListLocation() != null) {
+                validMetadataFilePaths.add(snapshot.manifestListLocation());
+            }
+
+            for (ManifestFile manifest : snapshot.allManifests(table.io())) {
+                if (!validMetadataFilePaths.add(manifest.path())) {
+                    // Already read this manifest
+                    continue;
+                }
+
+                try (ManifestReader<? extends ContentFile<?>> manifestReader = readerForManifest(table, manifest)) {
+                    for (ContentFile<?> contentFile : manifestReader) {
```

Comment:
This uses  `CloseableIterator` implicitly, and the close may be missing.

I don't think it should be "legal" to upcast `CloseableIterable` to `Iterable`.
or  `CloseableIterator` to  `Iterator`

would closing `ManifestReader` close the provided iterator's resources?
i.e. can we assume that the `CloseableIterator` implicitly created by the `for` loop won't have anything more to close than what `ManifestReader` already has?
but then, why would it be `CloseableIterator` at all, and not just `Iterator`?

Sorry, it seems i don't have yet basic understanding how these interfaces should or should not be used.
@rdblue @danielcweeks maybe you want to shed more light here? thanks!

Reply:
> I don't think it should be "legal" to upcast CloseableIterable to Iterable. or CloseableIterator to Iterator

Normally, yes. I think you're right.

In this case though, it looks like the ManifestReader is a `CloseableGroup` which takes ownership of closing any I/O it opens. 

> but then, why would it be CloseableIterator at all, and not just Iterator?

I don't have a good answer for that